resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java

```diff
@@ -830,49 +972,166 @@ void add(DnsRecord r) {
-                    return serverName;
+                if (serverName.address == null) {
+                    // Try to resolve via cache
+
+                    List<? extends DnsCacheEntry> entries = cache.get(serverName.nsName, null);
+                    if (entries == null || entries.isEmpty()) {
+                        // These will be resolved on the fly if needed.
+                        serverName.address = InetSocketAddress.createUnresolved(
+                                serverName.nsName, DefaultDnsServerAddressStreamProvider.DNS_PORT);
+                    } else {
+                        // We have some entries in the cache.
+                        for (int i = 0; i < entries.size(); i++) {
+                            DnsCacheEntry entry = entries.get(i);
+                            InetAddress address = entry.address();
+                            if (address != null) {
+                                if (serverName.address != null) {
+                                    AuthoritativeNameServer server = new AuthoritativeNameServer(serverName);
+                                    server.next = serverName.next;
+                                    serverName.next = server;
+                                    serverName = server;
+
+                                    count++;
+                                }
+                                // Use a TTL of Long.MIN_VALUE as we don't want to cache the actual address later on as
+                                // we have no idea about how long the TTL should be. When we encounter Long.MIN_VALUE
+                                // later on we will cache the unresolved address.
+                                serverName.update(parent.newRedirectServerAddress(address), Long.MIN_VALUE);
+                            }
+                        }
+                    }
                 }
                 serverName = serverName.next;
             }
-            return null;
         }
 
-        int size() {
-            return count;
+        /**
+         * Returns {@code true} if empty, {@code false} otherwise.
+         */
+        boolean isEmpty() {
+            return count == 0;
+        }
+
+        /**
+         * Creates a new {@link List} which holds the {@link InetSocketAddress}es.
+         */
+        List<InetSocketAddress> addressList() {
+            List<InetSocketAddress> addressList = new ArrayList<InetSocketAddress>(count);
+
+            AuthoritativeNameServer server = head;
+            while (server != null) {
+                if (server.address != null) {
+                    addressList.add(server.address);
+                }
+                server = server.next;
+            }
+            return addressList;
+        }
+
+        /**
+         * Add an authoritative nameserver to the cache if its not a root server.
+         */
+        void cache(AuthoritativeDnsServerCache cache, InetSocketAddress address, EventLoop loop) {
```

Comment:
I think we may have cross some wires in our offline discussion ... I think there are 2 parts to this puzzle.

#### Part 1 (cache insertion order)
my assumption was that we either cache insertion order is authority order, or we make a "best effort" to preserve server authority order if it avoids additional iteration/computation/memory. I don't think we want/need to preserve the order which the `uncachedRedirectDnsServerStream` method defines. If the user wants a custom order of elements in the cache, they can provide a custom cache implementation.

lets assume the following dns response:
```
AUTHORITY
foo.com NS ns1.foo.com.
foo.com NS ns2.foo.com.
foo.com NS ns3.foo.com.
foo.com NS ns4.foo.com.

ADDITIONAL
 ns2.foo.com. IPa
 ns4.foo.com. IPb
```

either the order we put into the cache is (server authority order):

```
ns1.foo.com. <unresovled, assuming not already in the cache>
ns2.foo.com IPa
ns3.foo.com. <unresovled, assuming not already in the cache>
ns4.foo.com IPb
```

or (best effort server authority order, preference for entries with additionals):

```
ns2.foo.com IPa
ns4.foo.com IPb
ns1.foo.com. <unresovled, assuming not already in the cache>
ns3.foo.com. <unresovled, assuming not already in the cache>
```

The `assuming not already in the cache` part is because we look up entries in the local cache which do not have any any additional entry to provide an IP (to try to cut down on DNS queries if we already know the answer). In `handleAll` `Long.MIN_VALUE` is currently used as a special value to treat differently in this case, but I don't think we need/want to modify the cache at all in this case.
The second option lines up with our algorithm/data-structures a bit more, so it maybe preferable.

#### Part 2 (ordering of redirect servers to query):
We can apply preference ordering to the redirect results be default, but can we clarify the `uncachedRedirectDnsServerStream` method name? This method name is confusing because we are actually caching results and entries maybe cached within. For example `newRedirectDnsServerStream` (with appropriate javadocs) would help avoid this confusion.

Reply:
ok got it... this will also make things easier :)

Reply:
For 1) I think we want to preserve "server authority order". 
 
For 2) renaming the method sounds good.

Reply:
done ... I wonder why IntelliJ did not highlight it :(